File: Codechef/Oct/MSNG.py
from collections import defaultdict

MAX_BASE = 37

def val(c): 
    if c >= '0' and c <= '9': 
        return ord(c) - ord('0') 
    else: 
        return ord(c) - ord('A') + 10; 
  
def toDeci(str,base): 
    llen = len(str) 
    power = 1
    num = 0
  
    for i in range(llen - 1, -1, -1): 
        if val(str[i]) >= base:
            return -1
        num += val(str[i]) * power 
        power = power * base 
    return num

def findAns(code, n):
    mp = defaultdict(int)
    for a, b in code.items():
        for i in set(b):
            mp[i] += 1
    # print mp
    for a, b in sorted(mp.items()):
        if b == n and a <= 10**12:
            return a
    return -1

t = int(raw_input())
for i in range(0, t):
    n = int(raw_input())
    code = defaultdict(list)
    for j in range(0, n):
        a, b = raw_input().split()
        a = int(a)
        if a == -1:
            for k in range(2, MAX_BASE):
                tmp = toDeci(b, k)
                if tmp != -1:
                    code[j].append(tmp)
        else:
            code[j].append(toDeci(b, a))
    print findAns(code, n)

File: Codechef/Oct/MSNG.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <algorithm>
#include <cmath>
#include <queue>
#include <cstring>
#include <sstream>
#include <map>
#include <climits>
#include <set>

using namespace std;

int MAX_BASE = 36;

int val(char c){ 
    if (c >= '0' && c <= '9') 
        return (int)c - '0'; 
    else
        return (int)c - 'A' + 10; 
} 
  
long long toDeci(int base, string &str){ 
    int len = str.size(); 
    long long power = 1;
    long long num = 0;
    long long limit = 1000000000000;
    for (int i = len - 1; i >= 0; i--){ 
        if (val(str[i]) >= base){ 
           return -1; 
        } 
        num += val(str[i]) * power; 
        // if (num > limit){
        // 	return -1;
        // }
        power = power * base; 
    } 
    return num; 
} 
  
long long findAnswer(vector<vector<long long>> &code, int n){
	long long mins = 1000000000001;
	for (int k=2; k<MAX_BASE; k++){
		long long c = code[0][k];
		if (c == -1) continue;
		int i=1;
		bool found = false;
		for (; i<n; i++){
			found = false;
			for (int x=2; x<MAX_BASE; x++){
				if (code[i][x] == c){
					found = true;
					break;
				}
			}
			if (!found) break;
		}

		if ((found || n == 1) && i == n){
			mins = min(mins, c);
		}
	}
	
	return mins == 1000000000001 ? -1 : mins;
}

void print(vector<vector<long long>> &code){
	for (auto row : code){
		for (auto cell : row){
			cout<<cell<<" ";
		}
		cout<<endl;
	}
}

int main(){
	freopen("input.txt", "r", stdin);
	// freopen("marm_02.txt", "w", stdout);
	int t;
	cin>>t;
	while (t--){
		int n, a;
		string b;
		cin>>n;
		vector<vector<long long>> code(n, vector<long long>(MAX_BASE, -1));
		for (int i=0; i<n; i++){
			cin>>a>>b;
			if (a == -1){
				for (int k=2; k<MAX_BASE; k++){
					code[i][k] = toDeci(k, b);
				}
			}else{
				code[i][a] = toDeci(a, b);
			}
		}

		// print(code);
		cout<<findAnswer(code, n)<<endl;
	}
 return 0;
}